lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/BinarySpec.hs

```diff
@@ -203,3 +237,123 @@ spec = do
+
+        it "decode (encode address) === address" $ property $ \addr ->
+            let
+                bytes = (runPut $ putAddress addr)
+                _hex = convertToBase @ByteString @ByteString Base16 (BL.toStrict bytes)
+            in
+                counterexample ("bytes: " <> show bytes) $ do
+                let addr' = try' $ runGet getAddress bytes
+                addr' === (Right addr)
+
+        it "decode (encode tx) === (validate tx)" $ property $ \(SignedTx signedTx) ->
+            let
+                bytes = try' (runPut $ putSignedTx signedTx)
+                hex = convertToBase @ByteString @ByteString Base16 . BL.toStrict <$> bytes
+                numberOfIns = length . inputs . fst $ signedTx
+                numberOfOuts = length . outputs . fst $ signedTx
+            in
+                counterexample ("bytes: " <> show hex) $
+                tabulate "number of inputs" [show numberOfIns] $
+                counterexample (show (numberOfIns, numberOfOuts)) $ do
+                let tx' = try' . unMessage . runGet getMessage =<< bytes
+                labelTx signedTx $ tx' === validate signedTx
+  where
+    unMessage :: Message -> (Tx, [TxWitness])
+    unMessage m = case m of
+        Transaction stx -> stx
+        _ -> error "expected a Transaction message"
+
+    try' :: a -> Either String a
+    try' = unsafePerformIO . wrap
+
+    wrap :: a -> IO (Either String a)
+    wrap = fmap (either (Left . show) Right) . (try @SomeException) . evaluate
+
+    labelTx stx = label (either id (const "Right _") (validate stx))
+
+    validate :: (Tx, [TxWitness]) -> Either String (Tx, [TxWitness])
+    validate stx@(Tx ins outs, _wits) = const stx <$> do
+        unless (length ins < 256) $
+            Left "number of inputs must be lower than 256"
+        unless (length outs < 256) $
+            Left "number of outputs must be lower than 256"
+        forM_ ins $ \(TxIn _tx ix, _coin) ->
+            unless (ix < 256) $ Left "input index must be < 256"
+
+
+-- Only generating single addresses!
+instance Arbitrary Address where
+    arbitrary = Address . appendTag 3 <$> genFixed 32
+    shrink (Address addr) = Address . appendTag 3 <$> shrinkFixedBS (BS.tail addr)
+
+-- Observation:
+-- genFixed and shrinkFixed would be nice candidates for DerivingVia.
+-- e.g.
+-- deriving instance Arbitrary Address via (ByteStringOfLength @33)
+genFixed :: Int -> Gen BS.ByteString
+genFixed n = BS.pack <$> (vectorOf n arbitrary)
+
+shrinkFixedBS :: ByteString -> [ByteString]
+shrinkFixedBS bs = if bs /= zeros then [zeros] else []
+      where
+        len = BS.length bs
+        zeros = BS.pack (replicate len 0)
+
+instance Arbitrary (Hash "Tx") where
+    arbitrary = Hash <$> genFixed 32
+    shrink (Hash bytes) = Hash <$> shrinkFixedBS bytes
+
+instance Arbitrary Coin where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
+
+-- | Generates mostly valid input indexes, but a few invalid ones.
```

Comment:
That doesn't seem right. Generators should **not** generate invalid data-types...